Newsletters: add 303 (2024-05-17) (#1669)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2024-05-17-newsletter.md

- [LDK #2907][] extends `OnionMessage` handling to accept an optional `Responder` input and return an object `ResponseInstructions` that contains the `response` message and the `reply_path`. This change enables asynchronous onion messaging responses and opens the door to more complex response mechanisms.
  
- [BDK #1403][] updates the `bdk_electrum` crate to make use of new sync/full-scan structures introduced in [BDK #1413][], queryable `CheckPoint` linked list [BDK #1369][], and cheaply-clonable transactions in `Arc`  pointers [BDK #1373][]. This change improves performance on wallet sync and transaction data scan. It is also now an option to fetch `TxOut`s to allow for fee calculation on transactions received from an external wallet.

> - BIPs #1458][] assigns [BIP352][] to the [Silent Payments][topic silent payments] proposal by Josie Baker and Ruben Somsen for a protocol for reusable payment addresses that generates a unique on-chain address each time it is used. The BIP draft was first discussed in newsletter [#255][news255 bip352].

Reviewer: 1458 has recently been merged. I guess this was supposed to be about the merge rather than BIP number assignment?

Reviewer 2: Yes it is about the merge, I'll make the change.

Author: @jirijakes "BIPs xxx assigns BIPyyy to z" is how we've been phrasing this for a long time.  It made sense years ago when getting a number usually happened immediately before a BIP was merged.  It didn't really make sense when there started being long merge delays in the BIPs repo.  Now that merge delays are being minimized again, I think it's probably worth sticking with our convention.

Reviewer: Ah, I have not realized until now that “assigns” here means that the BIP process of inclusion has been been completed. I assumed it was literally that the number has been assigned :-) Thanks for the explanation.

---
title: 'Bitcoin Optech Newsletter #303'
permalink: /en/newsletters/2024/05/17/
name: 2024-05-17-newsletter
slug: 2024-05-17-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME:harding

## News

- **Anonymous usage tokens:** Adam Gibson [posted][gibson autct] to
  Delving Bitcoin about a scheme he has developed to allow anyone who
  can [keypath][topic taproot] spend a UTXO to prove they could spend it
  without revealing which UTXO it is.  This follows Gibson's previous
  work developing anti-sybil mechanisms [PODLE][news85 podle] (used in
  the Joinmarket [coinjoin][topic coinjoin] implementation) and
  [RIDDLE][news205 riddle].

>   One example use he describes is announcing LN channels.  Each LN node

Reviewer 3: ```suggestion
  One use he describes is announcing LN channels.  Each LN node
```

  announces its channels to other LN nodes so that they can find paths
  for routing funds across the network.  Much of that channel
  information is stored in memory and the announcements are often
  rebroadcast to ensure they reach as many nodes as possible.  If an
  attacker could cheaply announce fake channels, they could waste an
  excessive amount of honest node memory and bandwidth in addition to
  disrupting pathfinding.  LN nodes deal with this today by
  only accepting announcements that are signed by a key that belongs to
  a valid UTXO.  <!-- ignoring for this description that the current LN

>   protocol requires only P2WPKH UTXOs --> That requires the channel

Reviewer 4: Invisible comment, but still:
```suggestion
  protocol requires only P2WSH UTXOs --> That requires the channel
```

  co-owners identify the specific UTXO they co-own, which may associate
  those funds with other past or future onchain transactions they create

>   (or leading to someone making an inaccurate association).

Reviewer 3: ```suggestion
  (or lead to someone making an inaccurate association).
```

  With Gibson's scheme, called anonymous usage tokens [with] curve trees
  (autct), the channel co-owners could sign a message without revealing
  their UTXO.  An attacker without a UTXO couldn't create a valid
  signature.  An attacker who did have a UTXO could create a
  valid signature---but they would have to keep as much money in that
  UTXO as an LN node would need to keep in a channel, limiting
  the worst case of any attack.  See [Newsletter #261][news261 lngossip]

>   for previous discussion of disassociating [channel

Reviewer 3: ```suggestion
  for a previous discussion of disassociating [channel
```

  announcements][topic channel announcements] from particular UTXOs.

  Gibson also describes several other ways autct could be used.  A
  basic mechanism for accomplishing this type of privacy---ring
  signatures---has been known for a long time, but Gibson uses a new
  cryptographic construction ([curve trees][]) to make the proofs more
  compact and faster to verify.  He also has each proof privately commit
  to the key used so a single UTXO can't be used to create an unlimited
  number of valid signatures.

  In addition to publishing [code][autct repo], Gibson also published a

>   proof-of-concept [forum][hodlboard] that requires providing a autct

Reviewer 3: ```suggestion
  proof-of-concept [forum][hodlboard] that requires providing an autct
```

>   proof in order to sign up, providing an environment where everyone is

Reviewer 3: ```suggestion
  proof to sign up, providing an environment where everyone is
```

  Andrew Poelstra, co-author of codex32, [replied][poelstra penlock1]
  with several comments and suggestions.  Without us trying both
  schemes---which would take several hours each---the exact set of
  tradeoffs between them isn't clear to us.  However, both seem to offer
  the same fundamental capabilities: instructions for securely
  generating a seed offline; the ability to split the seed into multiple
  shares using [Shamir's secret sharing][sss]; the ability to
  reconstitute the shares into the original seed; and the ability to
  verify checksums on both the shares and the original seed, allowing
  users to detect data corruption early when the original data might
  still be recoverable.

- **Alternative to BitVM:** Sergio Demian Lerner and several co-authors
  [posted][lerner bitvmx] to the Bitcoin-Dev mailing list about a new
  virtual CPU architecture based in part on the ideas behind
  [BitVM][topic acc].  The goal of their project, BitVMX, is to be able
  to efficiently prove the proper execution of any program that can be
  compiled to run on an established CPU architecture, such as
  [RISC-V][].  Like BitVM, BitVMX does not require any consensus changes

>   but it does require one or more designated parties act as a trusted

Reviewer 3: ```suggestion
  but does require one or more designated parties act as a trusted
```

>   but it does require one or more designated parties act as a trusted

Reviewer 3: ```suggestion
  but it does require one or more designated parties to act as a trusted
```

  verifier.  That means multiple users interactively participating in a
  contract protocol can prevent any one (or more) of the parties from
  withdrawing money from the contract unless that party successfully
  executes an arbitrary program specified by the contract.

  Lerner links to a [paper][bitvmx paper] about BitVMX which compares it
  to the original BitVM (see [Newsletter #273][news273 bitvm]) and to
  the limited details available about follow-up projects from the
  original BitVM developers.  An accompanying [website][bitvmx website]
  provides additional information in a slightly less technical form.

- **Continued discussion about updating BIP2:** Mark "Murch" Erhardt
  [continued][erhardt bip2] the discussion on the Bitcoin-Dev mailing
  list about updating [BIP2][], which is the document that currently

>   describes the Bitcoin improvement proposals (BIP) process.  His email

Reviewer 4: ```suggestion
  describes the Bitcoin Improvement Proposals (BIP) process.  His email
```

Author: Leaving as-is.  Up until about a year ago, we did uppercase words before introducing an abbreviation for them.  However, when I did that for Mastering Bitcoin, my editor ended up changing them all to lowercase.  I thought about it for a bit and decided that there really is no good reason to uppercase the expanded forms, so I switched to that style here also.

>   describes a number of problems, suggests solutions for many of them,

Reviewer 3: ```suggestion
  describes several problems, suggests solutions for many of them,
```

  and solicits feedback on his suggestions as well as proposals for
  solutions to the remaining problems.  For previous discussion about
  updating BIP2, see [Newsletter #297][news297 bip2]. <!-- TODO
  (optional): harding to add topic page for something like
  "Specification management" to track BIP and BINANA meta discussion -->

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND v0.18.0-beta.rc2][] is a release candidate for the next major

>   version of this popular LN node.

Reviewer 4: ```suggestion
  version of this popular LN node implementation.
```

release candidates.*

- [LND v0.18.0-beta.rc2][] is a release candidate for the next major
  version of this popular LN node implementation.

## Notable code and documentation changes

_Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo],
[Bitcoin Inquisition][bitcoin inquisition repo], and [BINANAs][binana
repo]._

- [Core Lightning #7190][] adds an additional offset `chainlag` into the CLTV
  value calculation to correctly target the current block height instead of the
  sync height. Payments are now safe and allowed during the blockchain sync
  process.

- [LDK #2973][] implements support for `OnionMessenger` to intercept messages on
  behalf of offline peers. It generates events on message interception and on
  the peer's return to online status for forwarding. Users should maintain an

>   allow list to only store messages for relevant peers. This is a stepping stone

Reviewer 3: ```suggestion
  _allow list_ to only store messages for relevant peers. This is a stepping stone
```
(because this is a definition... otherwise, if the reader doesn't take it as a definition, it sort of reads like a typo, at least to me)

>   to supporting offline payments through `held_htlc_available` [BOLTs

Reviewer: I believe these are called async payments.

```suggestion
  to supporting async payments through `held_htlc_available` [BOLTs
```

  #989][], which would allow an LSP (Lightning Service Provider) to hold onto
  an [HTLC][topic htlc] until a user comes back online.

> - [LDK #2907][] extends `OnionMessage` handling to accept an optional

Author: @Gustavojfe just a note that I'm going to update this description.  I infer from your description that Alice sends an HTLC to Bob, Bob realizes Carol is offline and so holds the message until she comes back online, then Bob delivers the HTLC.  Wallets have tried that before, but it's problematic because not immediately settling the HTLC means every hop along the path has their liquidity reduced.  That looks like an attack.

What I believe this PR is actually describing is that Alice wants to pay Carol but doesn't know if Carol is online, so Alice sends an _onion message_ to Carol via Bob (an onion message is data only, no money); Bob realizes Carol is offline and so holds the message until she comes back online; then Bob delivers the onion message to Carol; Carol opens the onion message which says "Alice wants to pay you, reply when you're ready to accept the payment"; Carol replies and Alice sends the payment.

>   `Responder` input and return an object `ResponseInstructions` that contains
>   the `response` message and the `reply_path`. This change enables asynchronous

Reviewer: I would suggest not going into details of `ResponseInstructions` here because, given it's an enum, it is more complicated than that. 

```suggestion
  `Responder` input and return an object `ResponseInstructions` that indicates how
  the response to the message should be handled. This change enables asynchronous
```

  onion messaging responses and opens the door to more complex response
  mechanisms.

- [BDK #1403][] updates the `bdk_electrum` crate to make use of new
  sync/full-scan structures introduced in [BDK #1413][], queryable `CheckPoint`
  linked list [BDK #1369][], and cheaply-clonable transactions in `Arc`

>   pointers [BDK #1373][]. This change improves performance on wallet sync and

Reviewer 5: Maybe? 
```suggestion
  pointers [BDK #1373][]. This change improves performance of wallet sync and
```

  transaction data scan. It is also now an option to fetch `TxOut`s to allow for
  fee calculation on transactions received from an external wallet.

> - [BIPs #1458][] assigns [BIP352][] to the [Silent Payments][topic silent

Reviewer 5: ```suggestion
- [BIPs #1458][] assigns [BIP352][] to the [silent payments][topic silent
```

We dont have silent payments listed here, but this jogged my mind to link you to this:
https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md

Reviewer 2: thanks for the feedback and for sharing tips

>   payments] proposal by Josie Baker and Ruben Somsen for a protocol for reusable

Reviewer 5: I think we mostly avoid names in this section 
```suggestion
  payments] proposal for a protocol for reusable
```

Author: As background for @Gustavojfe: we avoid names in the Merges section because getting something merged is a collaborative effort between both the authors and reviewers.  For code projects, a merge of code also usually indicates that it represents the rough consensus opinion of the project's maintainers (this obviously doesn't apply to BIPs).  It would be a lot of effort to list all of the reviewers[1], so we also don't list authors in order to be (somewhat) fair.

We do list names in the News section because we want to attribute ideas and opinions to individuals, since they may be positions that don't have widespread support.

[1] Especially since sometimes reviewers are spread across multiple places, as is often the case for BIPs, where review happens on the mailing lists, on the BIP repos, and often through personal correspondence and small-group discussion.

>   payment addresses that generates a unique on-chain address each time it is

Reviewer 5: ```suggestion
  payment addresses that generates a unique onchain address each time it is
```

>   used. The BIP draft was first discussed in newsletter [#255][news255 bip352].

Reviewer 5: Convention 

```suggestion
  used. The BIP draft was first discussed in [Newsletter #255][news255 bip352].
```
